### PR TITLE
Save logits of prediction results for node classification tasks

### PR DESCRIPTION
**Pull request by isratnisa**
Resolves [Issue](https://sim.amazon.com/issues/P77395067) 

The `predict()` function in node- and edge-decoder returns the final prediction (maximum value of logits) for multiclass classification.  Users may require the full set of logits, which is currently only returned for multilabel models.

To address this issue, this PR adds a new function `predict_proba()` which is enabled when `return_proba` flag is set. User can control whether to retrieve the full set of predictions or only the maximum by toggling this flag.

### CODE REVIEW

File: python/graphstorm/model/node_decoder.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def predict(self, inputs):
         """
         logits = th.matmul(inputs, self.decoder)
         # TODO(zhengda) we need to extend this to return probability of the prediction.
-        return logits if self._multilabel else logits.argmax(dim=1)
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/node_decoder.py:71]: What do you mean by `output saving logic`? We should always return the full set of `logits` in my understanding.

File: python/graphstorm/model/node_decoder.py [hunk 2]
```diff
@@ -65,10 +65,33 @@ def forward(self, inputs):
...
+        Parameters
+        ----------
+        inputs : tensor
+            The input features
+
+        Returns
+        -------
+        Tensor : all the predicted results
         """
         logits = th.matmul(inputs, self.decoder)
```
Comment (sojiadeshina) [inline python/graphstorm/model/node_decoder.py:93]: technically, we shouldn't return just the logits here.
we should apply softmax if it's not multilabel and apply sigmoid if it is multilabel

Comment (sojiadeshina) [reply python/graphstorm/model/node_decoder.py:93]: the reason why is that logits themselves are unnormalized. customer would expect normalized probability estimates.

File: python/graphstorm/model/node_gnn.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def forward(self, blocks, node_feats, _, labels, input_nodes=None):
         # weighted addition to the total loss
         return pred_loss + alpha_l2norm * reg_loss
 
-    def predict(self, blocks, node_feats, _, input_nodes):
+    def predict(self, blocks, node_feats, _, input_nodes, return_proba):
```
Comment (thvasilo) [inline python/graphstorm/model/node_gnn.py:128]: Is this function inherited from one of the mixins (GSgnnModel, GSgnnNodeModelInterface)? If yes we should ensure the arguments are consistent and override at the top level and ensure all concrete implementations are consistent as well.

File: python/graphstorm/model/node_decoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/node_decoder.py:71]: Done

File: python/graphstorm/model/node_gnn.py (hunk 3 above)
Comment (isratnisa) [reply python/graphstorm/model/node_gnn.py:131]: Done

**Review by sojiadeshina — APPROVED** [review]
LGTM

File: python/graphstorm/model/edge_gnn.py [hunk 4]
```diff
@@ -147,11 +152,14 @@ def edge_mini_batch_gnn_predict(model, loader, return_label=False):
...
     return_label : bool
-        Whether or not to return labels.
+        Whether or not to return labels
+    return_proba : bool
+        Return all the predicted results when true otherwise only the maximum value
 
     Returns
     -------
-    Tensor : GNN prediction results.
+    Tensor : GNN prediction results. Return all the results when return_proba is true
```
Comment (classicsong) [inline python/graphstorm/model/edge_gnn.py:161]: Does it return two tensors?

File: python/graphstorm/model/node_decoder.py [hunk 5]
```diff
@@ -65,10 +65,33 @@ def forward(self, inputs):
...
+        ----------
+        inputs : tensor
+            The input features
+
+        Returns
+        -------
+        Tensor : maximum of the predicted results
+        """
+        logits = th.matmul(inputs, self.decoder)
+        return (th.sigmoid(logits) > .5).long() if self._multilabel else logits.argmax(dim=1)
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:79]: Does this ((th.sigmoid(logits)) come from sklearn?

File: python/graphstorm/model/node_decoder.py [hunk 6]
```diff
@@ -65,10 +65,33 @@ def forward(self, inputs):
...
+        """ Make prediction on input data.
+
+        Parameters
+        ----------
+        inputs : tensor
+            The input features
+
+        Returns
+        -------
+        Tensor : all the predicted results
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:91]:
```suggestion
        Tensor : all normalized predicted results
```

**Review by classicsong — CHANGES_REQUESTED** [review]
Two general problems:
1. You need to add an argument to allow turn return_proba to True in np/ep_trainer/infer.
2. There is no test in the code. YOU need to add unit tests to cover all your code.

File: python/graphstorm/model/edge_gnn.py (hunk 4 above)
Comment (isratnisa) [reply python/graphstorm/model/edge_gnn.py:161]: Yes, depending on the `return_label` and `return_proba` flag.

File: python/graphstorm/model/edge_decoder.py [hunk 7]
```diff
@@ -132,10 +132,36 @@ def predict(self, g, h):
...
             if self.regression:
                 out = self.regression_head(out)
-            elif not self._multilabel:
+            elif self._multilabel:
+                out = (th.sigmoid(out) > .5).long()
+            else:  # not multilabel
                 out = out.argmax(dim=1)
         return out
 
+    def predict_proba(self, g, h):
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_decoder.py:141]: Are you suggesting to get rid of `pred/pred_proba()` functions?

File: python/graphstorm/model/node_decoder.py (hunk 5 above)
Comment (isratnisa) [reply python/graphstorm/model/node_decoder.py:79]: We apply sigmoid to make the logits between (0, 1)

File: tests/unit-tests/test_gnn.py [hunk 8]
```diff
@@ -107,6 +107,12 @@ def check_node_prediction(model, data):
     assert_almost_equal(pred1[0:len(pred1)].numpy(), pred2[0:len(pred2)].numpy(), decimal=5)
     assert_equal(labels1.numpy(), labels2.numpy())
 
+    # Test the return_proba argument.
+    pred3, labels3 = node_mini_batch_predict(model, embs, dataloader1, return_proba=True, return_label=True)
+    assert pred3.dim() == 2  # returns all predictions (2D tensor) when return_proba is true
+    pred4, labels4 = node_mini_batch_predict(model, embs, dataloader1, return_proba=False, return_label=True)
+    assert(pred4.dim() == 1)  # returns maximum prediction (1D tensor) when return_proba is False
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:114]: Can we check the value of pred3 and pred4?

File: tests/unit-tests/test_gnn.py [hunk 9]
```diff
@@ -130,6 +136,12 @@ def check_mlp_node_prediction(model, data):
...
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:143]: Same here.

File: python/graphstorm/eval/eval_func.py [hunk 10]
```diff
@@ -207,6 +207,7 @@ def eval_roc_auc(logits,labels):
 def eval_acc(pred, labels):
     """compute evaluation accuracy.
     """
+    pred = pred.argmax(dim=1) if pred.dim() > 1 else pred
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/eval/eval_func.py]: Modified.

File: python/graphstorm/model/edge_decoder.py [hunk 11]
```diff
@@ -132,10 +132,41 @@ def predict(self, g, h):
...
+            ifeat = h[dest_type][v]
+            out = th.einsum('ai,bij,aj->ab', ufeat, self.basis_para.to(ifeat.device), ifeat)
+            out = self.combine_basis(out)
+            if self.regression:
+                out = self.regression_head(out)
+            elif self._multilabel:
+                out = th.sigmoid(out)
+            else:
+                out = th.softmax(out, 0)
+        return out
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_decoder.py:168]: Modified.

File: tests/unit-tests/test_gnn.py (hunk 8 above)
Comment (isratnisa) [reply tests/unit-tests/test_gnn.py:147]: Done.

File: tests/unit-tests/test_gnn.py (hunk 9 above)
Comment (isratnisa) [reply tests/unit-tests/test_gnn.py:179]: Done.

File: python/graphstorm/model/edge_decoder.py [hunk 12]
```diff
@@ -242,19 +273,24 @@ def predict(self, g, h):
...
+        g : DGLBlock
+            The minibatch graph
+        h : dict of Tensors
+            The dictionary containing the embeddings
+
+        Returns
+        -------
+        Tensor : the scores of each edge.
+        """
+        return self.predict(g,h)
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_decoder.py]: Resolved.

File: python/graphstorm/config/argument.py [hunk 13]
```diff
@@ -1562,6 +1574,10 @@ def _add_node_classification_args(parser):
             "The weights should be in the following format 0.1,0.2,0.3,0.1,0.0 ")
     group.add_argument("--num-classes", type=int, default=argparse.SUPPRESS,
                        help="The cardinality of labels in a classifiction task")
+    group.add_argument("--return-proba", type=bool, default=True,
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/config/argument.py]: Done

File: python/graphstorm/eval/eval_func.py [hunk 14]
```diff
@@ -207,6 +207,12 @@ def eval_roc_auc(logits,labels):
 def eval_acc(pred, labels):
     """compute evaluation accuracy.
     """
+    if pred.dim() > 1:
+        # if pred has dimension > 1, it has full logits instead of final prediction
+        assert th.is_floating_point(pred)
+        pred = pred.argmax(dim=1)
+    # Check if pred is integer tensor
+    assert(not th.is_floating_point(pred) and not th.is_complex(pred))
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/eval/eval_func.py]: Done

File: python/graphstorm/trainer/np_trainer.py [hunk 15]
```diff
@@ -50,7 +50,8 @@ def fit(self, train_loader, num_epochs,
             save_model_path=None,
             save_model_frequency=-1,
             save_perf_results_path=None,
-            freeze_input_layer_epochs=0):
+            freeze_input_layer_epochs=0,
+            return_proba=True):
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/trainer/np_trainer.py]: Removed

**Review by classicsong — APPROVED** [review]
LGTM

File: python/graphstorm/config/argument.py [hunk 16]
```diff
@@ -1522,6 +1533,9 @@ def _add_node_classification_args(parser):
...
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/config/argument.py]: Got it. There are multiple places where we are using just `type=bool`.  Keeping it for consistency. Will fix them all in a separate PR.

File: python/graphstorm/model/edge_gnn.py [hunk 17]
```diff
@@ -127,17 +130,19 @@ def forward(self, blocks, batch_graph, node_feats, _,
...
             # no GNN message passing in encoder
             encode_embs = self.comput_input_embed(input_nodes, node_feats)
         else:
             encode_embs = self.compute_embed_step(blocks, node_feats)
+        if return_proba:
+            return self.decoder.predict_proba(batch_graph, encode_embs)
         return self.decoder.predict(batch_graph, encode_embs)
 
-def edge_mini_batch_gnn_predict(model, loader, return_label=False):
+def edge_mini_batch_gnn_predict(model, loader, return_proba=True, return_label=False):
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_gnn.py:145]: Right! The reason I changed the order is to be consistent with the code base (most of the function arguments follow the pattern of `pred`->`label`. E.g., `def eval_acc(pred, labels)`), also the repo is relatively new. But you are right. From now on, we should add new arg with default value at the end.

File: python/graphstorm/model/node_gnn.py [hunk 18]
```diff
@@ -188,7 +197,7 @@ def node_mini_batch_gnn_predict(model, loader, return_label=False):
     else:
         return preds, embs
 
-def node_mini_batch_predict(model, emb, loader, return_label=False):
+def node_mini_batch_predict(model, emb, loader, return_proba=True, return_label=False):
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/node_gnn.py:200]: Ditto
